File: backend/app/routes/whiteboards.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import Optional

from ..auth import get_current_user
from ..database import get_db
from ..models import Whiteboard, WhiteboardElement, WhiteboardConnection
from ..schemas.whiteboard import (
    WhiteboardResponse,
    WhiteboardElementCreate,
    WhiteboardElementUpdate,
    WhiteboardElementResponse,
    WhiteboardConnectionCreate,
    WhiteboardConnectionUpdate,
    WhiteboardConnectionResponse,
)

router = APIRouter()


@router.get("/whiteboards", response_model=WhiteboardResponse)
def get_or_create_whiteboard(project_id: str, db: Session = Depends(get_db), user=Depends(get_current_user)):
    board = db.query(Whiteboard).filter(Whiteboard.project_id == project_id).first()
    if not board:
        board = Whiteboard(project_id=project_id)
        db.add(board)
        db.commit()
        db.refresh(board)
    return board


@router.get("/whiteboards/{board_id}", response_model=WhiteboardResponse)
def get_board(board_id: str, db: Session = Depends(get_db), user=Depends(get_current_user)):
    board = db.query(Whiteboard).filter(Whiteboard.id == board_id).first()
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    return board


@router.post("/whiteboards/{board_id}/elements", response_model=WhiteboardElementResponse)
def create_element(board_id: str, payload: WhiteboardElementCreate, db: Session = Depends(get_db), user=Depends(get_current_user)):
    board = db.query(Whiteboard).filter(Whiteboard.id == board_id).first()
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    el = WhiteboardElement(board_id=board_id, **payload.dict())
    db.add(el)
    db.commit()
    db.refresh(el)
    return el


@router.patch("/whiteboard-elements/{element_id}", response_model=WhiteboardElementResponse)
def update_element(element_id: str, payload: WhiteboardElementUpdate, db: Session = Depends(get_db), user=Depends(get_current_user)):
    try:
        el = db.query(WhiteboardElement).filter(WhiteboardElement.id == element_id).first()
        if not el:
            raise HTTPException(status_code=404, detail="Element not found")
        
        update_data = payload.dict(exclude_unset=True)
        allowed_fields = {'x', 'y', 'width', 'height', 'rotation', 'z_index', 'text', 'fill', 'text_color', 'font_family', 'font_size'}
        
        for key, value in update_data.items():
            if key in allowed_fields:
                setattr(el, key, value)
        
        db.commit()
        db.refresh(el)
        
        element_dict = {
            'id': el.id,
            'board_id': el.board_id,
            'type': el.type,
            'x': el.x,
            'y': el.y,
            'width': el.width,
            'height': el.height,
            'rotation': el.rotation,
            'z_index': el.z_index,
            'text': el.text,
            'fill': el.fill,
            'text_color': el.text_color,
            'font_family': el.font_family,
            'font_size': el.font_size,
            'created_at': el.created_at,
            'updated_at': el.updated_at
        }
        return WhiteboardElementResponse(**element_dict)
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        print(f"Error updating whiteboard element: {e}")
        import traceback
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"Error updating element: {str(e)}")


@router.delete("/whiteboard-elements/{element_id}")
def delete_element(element_id: str, db: Session = Depends(get_db), user=Depends(get_current_user)):
    el = db.query(WhiteboardElement).filter(WhiteboardElement.id == element_id).first()
    if not el:
        raise HTTPException(status_code=404, detail="Element not found")
    db.delete(el)
    db.commit()
    return {"ok": True}


@router.post("/whiteboards/{board_id}/connections", response_model=WhiteboardConnectionResponse)
def create_connection(board_id: str, payload: WhiteboardConnectionCreate, db: Session = Depends(get_db), user=Depends(get_current_user)):
    board = db.query(Whiteboard).filter(Whiteboard.id == board_id).first()
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    conn = WhiteboardConnection(board_id=board_id, **payload.dict())
    db.add(conn)
    db.commit()
    db.refresh(conn)
    return conn


@router.patch("/whiteboard-connections/{connection_id}", response_model=WhiteboardConnectionResponse)
def update_connection(connection_id: str, payload: WhiteboardConnectionUpdate, db: Session = Depends(get_db), user=Depends(get_current_user)):
    conn = db.query(WhiteboardConnection).filter(WhiteboardConnection.id == connection_id).first()
    if not conn:
        raise HTTPException(status_code=404, detail="Connection not found")
    for key, value in payload.dict(exclude_unset=True).items():
        setattr(conn, key, value)
    db.commit()
    db.refresh(conn)
    return conn


@router.delete("/whiteboard-connections/{connection_id}")
def delete_connection(connection_id: str, db: Session = Depends(get_db), user=Depends(get_current_user)):
    conn = db.query(WhiteboardConnection).filter(WhiteboardConnection.id == connection_id).first()
    if not conn:
        raise HTTPException(status_code=404, detail="Connection not found")
    db.delete(conn)
    db.commit()
    return {"ok": True}

File: front/src/api/whiteboards.ts
import { apiClient } from './client';
import type { Shape, SectionShape, Comment } from '../components/whiteboard/types';

export type WhiteboardElementDto = {
  id: string;
  board_id: string;
  type: string;
  x: number;
  y: number;
  width: number;
  height: number;
  rotation: number;
  z_index: number;
  text?: string | null;
  fill?: string | null;
  text_color?: string | null;
  font_family?: string | null;
  font_size: number;
  task_id?: string | null;
};

export type WhiteboardConnectionDto = {
  id: string;
  board_id: string;
  from_element_id?: string | null;
  to_element_id?: string | null;
  stroke?: string | null;
  stroke_width?: number | null;
  points?: string | null;
};

export type WhiteboardDto = {
  id: string;
  project_id: string;
  elements: WhiteboardElementDto[];
  connections: WhiteboardConnectionDto[];
};

function normalizeElement(raw: any): WhiteboardElementDto {
  return {
    id: raw.id,
    board_id: raw.board_id ?? raw.boardId,
    type: raw.type,
    x: raw.x,
    y: raw.y,
    width: raw.width,
    height: raw.height,
    rotation: raw.rotation ?? 0,
    z_index: raw.z_index ?? raw.zIndex ?? 0,
    text: raw.text ?? null,
    fill: raw.fill ?? null,
    text_color: raw.text_color ?? raw.textColor ?? null,
    font_family: raw.font_family ?? raw.fontFamily ?? null,
    font_size: raw.font_size ?? raw.fontSize ?? 14,
    task_id: raw.task_id ?? raw.taskId ?? null,
  };
}

function normalizeConnection(raw: any): WhiteboardConnectionDto {
  return {
    id: raw.id,
    board_id: raw.board_id ?? raw.boardId,
    from_element_id: raw.from_element_id ?? raw.fromElementId ?? null,
    to_element_id: raw.to_element_id ?? raw.toElementId ?? null,
    stroke: raw.stroke ?? null,
    stroke_width: raw.stroke_width ?? raw.strokeWidth ?? null,
    points: raw.points ?? null,
  };
}

export async function getOrCreateWhiteboard(projectId: string): Promise<WhiteboardDto> {
  const { data } = await apiClient.get('/whiteboards', {
    params: { project_id: projectId },
  });
  return {
    id: data.id,
    project_id: data.project_id ?? data.projectId,
    elements: Array.isArray(data.elements) ? data.elements.map(normalizeElement) : [],
    connections: Array.isArray(data.connections) ? data.connections.map(normalizeConnection) : [],
  };
}

export async function createElement(
  boardId: string,
  element: {
    type: string;
    x: number;
    y: number;
    width: number;
    height: number;
    rotation?: number;
    z_index?: number;
    text?: string;
    fill?: string;
    text_color?: string;
    font_family?: string;
    font_size?: number;
    task_id?: string;
  }
): Promise<WhiteboardElementDto> {
  const { data } = await apiClient.post(`/whiteboards/${boardId}/elements`, element);
  return normalizeElement(data);
}

export async function updateElement(
  elementId: string,
  updates: {
    type?: string;
    x?: number;
    y?: number;
    width?: number;
    height?: number;
    rotation?: number;
    z_index?: number;
    text?: string;
    fill?: string;
    text_color?: string;
    font_family?: string;
    font_size?: number;
    task_id?: string;
  }
): Promise<WhiteboardElementDto> {
  const { data } = await apiClient.patch(`/whiteboard-elements/${elementId}`, updates);
  return normalizeElement(data);
}

export async function deleteElement(elementId: string): Promise<void> {
  await apiClient.delete(`/whiteboard-elements/${elementId}`);
}

export async function createConnection(
  boardId: string,
  connection: {
    from_element_id?: string;
    to_element_id?: string;
    stroke?: string;
    stroke_width?: number;
    points?: string;
  }
): Promise<WhiteboardConnectionDto> {
  const { data } = await apiClient.post(`/whiteboards/${boardId}/connections`, connection);
  return normalizeConnection(data);
}

export async function updateConnection(
  connectionId: string,
  updates: {
    from_element_id?: string;
    to_element_id?: string;
    stroke?: string;
    stroke_width?: number;
    points?: string;
  }
): Promise<WhiteboardConnectionDto> {
  const { data } = await apiClient.patch(`/whiteboard-connections/${connectionId}`, updates);
  return normalizeConnection(data);
}

export async function deleteConnection(connectionId: string): Promise<void> {
  await apiClient.delete(`/whiteboard-connections/${connectionId}`);
}

export async function getProjectSections(projectId: string): Promise<SectionShape[]> {
  const { data } = await apiClient.get('/whiteboards/sections', {
    params: { project_id: projectId },
  });
  return Array.isArray(data) ? data : [];
}

export async function linkElementToTask(elementId: string, taskId: string): Promise<WhiteboardElementDto> {
  const { data } = await apiClient.patch(`/whiteboard-elements/${elementId}/link-task`, { task_id: taskId });
  return normalizeElement(data);
}

export async function unlinkElementFromTask(elementId: string): Promise<WhiteboardElementDto> {
  const { data } = await apiClient.patch(`/whiteboard-elements/${elementId}/unlink-task`, {});
  return normalizeElement(data);
}

export async function getProjectTasks(projectId: string) {
  const { data } = await apiClient.get(`/projects/${projectId}/tasks`);
  return Array.isArray(data) ? data : [];
}
